Add adanovo v1

BioGeek opened the pull request:
This PR adds the adanovo v1 algorithm.

What still needs to be done:
* [download a checkpoint with pretrained weights](https://github.com/PominovaMS/denovo_benchmarks/blob/b3a2c77c284f6b58dd189ce9c9c4525d9d40f3df/algorithms/adanovo/container.def#L36)
* [Update `algorithms/adanovo/make_predictions.sh` to use those pretrained weights](https://github.com/PominovaMS/denovo_benchmarks/blob/b3a2c77c284f6b58dd189ce9c9c4525d9d40f3df/algorithms/adanovo/make_predictions.sh#L18)
* add an `output_mapper.py` file to convert the adanovo output to the common output format.

@TreeNode147 & @junxia97: you can add the above modifications directly to this PR.

TreeNode147 commented:
Hello, I tried, but it seems I don't have the permissions to modify this branch. Please directly update my modifications into the branch.

1. container.def

```
Bootstrap: docker
# Define the base image to inherit from.
# (e.g. image with a particular python version
# or a particular pytorch/tensorflow version).
From: python:3.10

# Define system variables to provide GPU access within the container.
%environment
    export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH

%files
    # Copy algorithm-related files to a separate dir /algo.
    # Don't change the dir name.
    algorithms/adanovo /algo
    algorithms/base /algo/base

%post
	# [Optional] Install system packages
	# (e.g. some base images may need git installation)
		
    # [Optional] Download algorithm-related files
    # (source codes, weights, etc.)
    # All files must be placed within /algo dir.
    cd /algo
    git clone https://github.com/Westlake-OmicsAI/adanovo_v1.git
    
    # [Optional] Install dependencies
    # (pandas is recommended to support parsing dataset tags)
    pip install --no-cache-dir pandas
    # install or build from source the algorithm, etc.
    cd adanovo_v1
    pip install -r requirements.txt

    # Download model weights
    # cd /algo
    # TODO: add model weights download link
    curl -L -O https://github.com/Westlake-OmicsAI/adanovo_v1/releases/download/add_a_ckpt/ckpt_1.ckpt
    
%post
	# Make sure make_predictions.sh file is executable.
    chmod +x /algo/make_predictions.sh
    
# Run algorithm and convert outputs.
# Data is expected to be mounted into /algo/data dir.
%runscript
    cd /algo && ./make_predictions.sh data

```

2. make_predictions.sh

```
#!/bin/bash

# Get dataset property tags
DSET_TAGS=$(python /algo/base/dataset_tags_parser.py --dataset "$@")
# Parse tags and set individual environment variables for each of them
# (variable names are identical to tag names
#  -- check DatasetTag values in dataset_config.py)
while IFS='=' read -r key value; do
    export "$key"="$value"
done <<< "$DSET_TAGS"

# Iterate through files in the dataset
for input_file in "$@"/*.mgf; do

    echo "Processing file: $input_file"

    # Convert input data to model format
    python ./adanovo_v1/adanovo.py --mode=denovo --model=./ckpt_1.ckpt --peak_path="$input_file" --config=./adanovo_v1/config.yaml --output=./demo.mztab

    # Convert predictions to the general output format
    python ./output_mapper.py --output_path=./demo.mztab --input_dir="$@"
done
```

3. output_mapper.py

```
"""
Script to convert predictions from the algorithm output format 
to the common output format.
"""

import argparse
import os
import re
from tqdm import tqdm
from pyteomics import mgf
from pyteomics.mztab import MzTab
from base import OutputMapperBase


class OutputMapper(OutputMapperBase):
    REPLACEMENTS = [
        ("C+57.021", "C[UNIMOD:4]"),
        # Amino acid modifications.
        ("M+15.995", "M[UNIMOD:35]"),    # Met oxidation
        ("N+0.984", "N[UNIMOD:7]"),     # Asn deamidation
        ("Q+0.984", "Q[UNIMOD:7]"),     # Gln deamidation
        # N-terminal modifications.
        ("+42.011", "[UNIMOD:1]"),      # Acetylation
        ("+43.006", "[UNIMOD:5]"),      # Carbamylation
        ("-17.027", "[UNIMOD:385]"),     # NH3 loss
        # "+43.006-17.027": 25.980265      # Carbamylation and NH3 loss
    ]
    N_TERM_MOD_PATTERN = r"^(\[UNIMOD:[0-9]+\])" # find N-term modifications

    def __init__(self, input_dir: str) -> None:
        """TODO."""
        fnames = [fname for fname in os.listdir(input_dir) if fname.endswith(".mgf")]
        self.file_names = [fname.split(".")[0] for fname in sorted(fnames)]

        file_paths = [
            os.path.join(input_dir, fname + ".mgf")
            for fname in self.file_names
        ]
        file_paths = sorted(file_paths)
        return
    
    def _transform_match_n_term_mod(self, match: re.Match) -> str:
        """
        Transform representation of peptide substring matching
        the N-term modification pattern.
        `[n_mod]PEP` -> `[n_mod]-PEP`
        
        Parameters
        ----------
        match : re.Match
            Substring matching the N-term modification pattern.

        Returns
        -------
        transformed_match : str
            Transformed N-term modification pattern representation.
        """
        ptm = match.group(1)
        return f"{ptm}-"

    def _spectrum_id_to_filename_idx(self, spectrum_id: str) -> str:
        file_idx, spectrum_idx = spectrum_id.split(":")
        
        file_idx = int(file_idx) - 1
        filename = self.file_names[file_idx]        
        return filename, spectrum_idx
    
    def _parse_scores(self, scores: str) -> list[float]:
        """
        Convert per-token scores from a string of float scores 
        separated by ',' to a list of float numbers.
        """
        scores = scores.split(",")
        scores = list(map(float, scores))
        return scores

    def format_spectrum_id(self, spectrum_id: str) -> str:
        """
        TODO.
        Transform scan index generated by the algorithm to the common format
        `ms_run[i]:index=j` -> `filename:j`.
        """
        spectrum_id = re.sub("[a-z=_\[\]]", "", spectrum_id)

        filename, spectrum_idx = self._spectrum_id_to_filename_idx(spectrum_id)
        spectrum_id = filename + ":" + spectrum_idx
        return spectrum_id
    
    def format_sequence(self, sequence: str) -> str:
        """
        Convert peptide sequence to the common output data format 
        (ProForma with modifications represented 
        in the delta mass notation).

        Parameters
        ----------
        sequence : str
            Peptide sequence in the original algorithm output format.

        Returns
        -------
        transformed_sequence : str
            Peptide sequence in the common output data format.   
        """

        # direct (token-to-token) replacements
        for repl_args in self.REPLACEMENTS:
            sequence = sequence.replace(*repl_args)

        # transform n-term modification notation
        # represent in ProForma delta mass notation [+n_term_mod]-PEP
        if re.search(self.N_TERM_MOD_PATTERN, sequence):
            sequence = re.sub(self.N_TERM_MOD_PATTERN, self._transform_match_n_term_mod, sequence)

        return sequence


parser = argparse.ArgumentParser()
parser.add_argument(
    "--output_path", required=True, help="The path to the algorithm predictions file."
)
parser.add_argument(
    "--input_dir", required=True, help="The dir with the input dataset .mgf files.",
)
args = parser.parse_args()

# Read predictions from output file
output_data = MzTab(args.output_path)
output_data = output_data.spectrum_match_table

# Rename columns to the expected column names if needed
output_data = output_data.rename(
    {
        "search_engine_score[1]": "score",
        "spectra_ref": "spectrum_id",
        "opt_ms_run[1]_aa_scores": "aa_scores",
    },
    axis=1,
)

# Transform data to the common output format
output_mapper = OutputMapper(input_dir=args.input_dir)
output_data = output_mapper.format_output(output_data)

# Save processed predictions to outputs.csv
# (the expected name for the algorithm output file)
output_data.to_csv("outputs.csv", index=False)

```

In the [/algorithms/adanovo](url) directory, there are only these three files, and nothing else.

The checkpoint will be downloaded in the container.def, and you can also find it at the link: 
[https://github.com/Westlake-OmicsAI/adanovo_v1/releases/download/add_a_ckpt/ckpt_1.ckpt](url)

BioGeek commented:
Hi @TreeNode147 

I have added the code you provided in [e82cd5d](https://github.com/PominovaMS/denovo_benchmarks/pull/38/commits/e82cd5d22a0c077f485deca1960694d901877eb9). Thanks!

When I tried to run the code, I got the error

```bash
Error: Invalid value for '--model': File './ckpt_1.ckpt' does not exist.
```

So In [77a61bc](https://github.com/PominovaMS/denovo_benchmarks/pull/38/commits/77a61bcc8f0fd85874453cf10ac180f4cf995a9b) I made a small modification to put the checkpoint at the correct location inside the `/algo` folder. 

I have tested this on the [`sample_data/9_species_human/mgf/151009_exo4_1.mgf`](https://github.com/PominovaMS/denovo_benchmarks/blob/main/sample_data/9_species_human/mgf/151009_exo4_1.mgf) sample file and am now able to get a `demo.mztab` file with predictions (see `demo.mztab` in [adanovo_outputs.zip](https://github.com/user-attachments/files/17617258/adanovo_outputs.zip)).

However, I noticed that in the `outputs.csv` , the  post-translational modifications weren't correctly replaced with their equivalent Unimod accession codes. The output contained lines like:

```
sequence,PSM_ID,accession,unique,database,database_version,search_engine,score,modifications,retention_time,charge,exp_mass_to_charge,calc_mass_to_charge,spectrum_id,pre,post,start,end,aa_scores
QLEEC(+57.02)AEALFK,9,,,,,"[MS, MS:1003281, Adanovo]",-0.4298327323049307,,,2.0,667.17725,669.32392831688,151009_exo4_1:8,,,,,"0.59790,0.32262,0.55578,0.50398,0.65238,0.66080,0.64955,0.39343,0.44902,0.63617,0.63530"
GSDENN(+.98)GKPEEDASEEDNPGHSQHHEER,240,,,,,"[MS, MS:1003281, Adanovo]",-0.19280607124854776,,,4.0,783.0948,783.5664883918802,151009_exo4_1:239,,,,,"0.77969,0.46862,0.82158,0.83856,0.49767,0.89566,0.87091,0.57719,0.87662,0.90197,0.89120,0.89864,0.86482,0.65728,0.89270,0.89957,0.89662,0.90268,0.80471,0.85526,0.58731,0.78985,0.88758,0.55126,0.89664,0.89882,0.89823,0.90340"
```

Notice the `C(+57.02)` and `N(+.98)`. See `outputs_original.csv` in the attachment for the full file.

In [ebbfe87](https://github.com/PominovaMS/denovo_benchmarks/pull/38/commits/ebbfe87fb39d4adb6bb756ad0aa5491c158f7113) I have modified `output_mapper.py` and now the PTMs seem to be correcly replaced, see `outputs_unimod.csv` in the attachement.

TreeNode147 commented:
OK, thank you very much for your modifications and suggestions.
